internal/agent/upgrade.go

```diff
@@ -148,17 +95,133 @@ func determineUpgradeImage(version string) (string, error) {
 		Version: version,
 	})
 	if err != nil {
-		return "", err
+		return nil, err
 	}
 
 	if img != "" {
-		return img, nil
+		return nil, nil
 	}
 
 	registry, err := utils.OSRelease("IMAGE_REPO")
 	if err != nil {
-		return "", fmt.Errorf("can't find IMAGE_REPO key under /etc/os-release %w", err)
+		return nil, fmt.Errorf("can't find IMAGE_REPO key under /etc/os-release %w", err)
+	}
+
+	return v1.NewSrcFromURI(fmt.Sprintf("%s:%s", registry, version))
+}
+
+// generateUpgradeConfForCLIArgs creates a kairos configuration for `--source` and `--recovery`
+// command line arguments. It will be added to the rest of the configurations.
+func generateUpgradeConfForCLIArgs(source string, upgradeRecovery bool) (string, error) {
+	upgrade := map[string](map[string]interface{}){
+		"upgrade": {},
+	}
+
+	if upgradeRecovery {
+		upgrade["upgrade"]["recovery"] = "true"
+	}
+
+	// Set uri both for active and recovery because we don't know what we are
+	// actually upgrading. The "upgradeRecovery" is just the command line argument.
+	// The user might have set it to "true" in the kairos config. Since we don't
+	// have access to that yet, we just set both uri values which shouldn't matter
+	// anyway, the right one will be used later in the process.
+	if source != "" {
+		upgrade["upgrade"]["recovery-system"] = map[string]string{
+			"uri": source,
```

Comment:
```
if upgradeRecovery{
  entry = "recovery-system"
else {
  entry = "system"
}

upgrade["upgrade"][entry] = map[string]string{
			"uri": source,
}
```

in fact, it can be even easier if we just declare the entry var at the start with the `entry := "system"` value and on line 120 we override it when we enter the `if upgradeRecovery`

Then in here we just set the `upgrade["upgrade"][entry] = map[string]string{"uri": source}` if `source != ""`

Reply:
The problem described in the comment still remains. What if in the kairos config we have set `upgrade.recovery = true` ? We don't know this at this point because we haven't read the config yet so we can't simply rely on the command line value for upgradeRecovery because that's only one way to set this.

Reply:
Discussed in a call, leaving as is.